### PR TITLE
Color identifiers mode

### PR DESCRIPTION
**Pull request by ralesi**
Add color identifiers mode, toggled by `SPC t C i`.

### CODE REVIEW

**Comment by mbertheau** [comment]
Hmm, how does this play with rainbow-identifiers, which is on the same key binding?https://github.com/syl20bnr/spacemacs/blob/b8f3d395fd2ac6b969d4b19cc47d37663056803a/layers/colors/packages.el#L46

**Comment by ralesi** [comment]
Oops,  I noticed it free but I didnt load all packages before doing so.

**Comment by syl20bnr** [comment]
What are the difference with `rainbow-identifier` (in the colors layer) ?

**Comment by ralesi** [comment]
rainbow-identifiers seems to color all unique text in the buffer with different colors.  color-identifiers-mode on the other hand, recognizes variable in the source and colors only the variables in the source.  I should add this to the color package though.  Will do that now...

File: layers/colors/packages.el [hunk 1]
```diff
@@ -23,6 +24,21 @@
 ;;     :disabled t
 ;;     :init (add-hook 'emacs-lisp-mode-hook 'rainbow-blocks-mode)))
 
+(defun colors/init-color-identifiers-mode ()
+  (use-package color-identifiers-mode
+    :commands color-identifiers-mode
+    :defer t
+    :init
+    (progn
+      (spacemacs|diminish color-identifiers-mode "C " "C ")
```
Comment (mbertheau) [inline layers/colors/packages.el]: The first C has got to be Ⓒ (compare https://github.com/syl20bnr/spacemacs/blob/d993a021847cde2c42865bab6afa8adbb2edda0b/spacemacs/packages.el#L221)

**Comment by mbertheau** [comment]
I think we should form an opinion which identifier coloring package is better and include only one by default. @ralesi's description above sounds like color-identifiers-mode is more sensible than rainbow-identifiers.

To that end I tried this PR locally, but toggling it with `tCI` doesn't change identifier color from their bland gray. color-identifiers mode is active. Does somebody know how to debug this?

**Comment by ralesi** [comment]
@mbertheau, this mode is relatively unobtrusive, it will not change the text for common patterns, unless it recognizes it as a variable.  Look at a `let` or `setq` construct and see if you can see the colors change then.

**Comment by syl20bnr** [comment]
@ralesi can you make a screenshot with rainbow identifier as well so we can compare ?

I see an advantage of `color-indentifiers`, it tries to be smart about the choice of colors, I have to test with different themes to see how it looks. In `rainbow-identifiers` the saturation and lightness have to be adjusted for each theme (a couple are preconfigured in the layer).

**Comment by ralesi** [comment]
Here is the same section with `rainbow-identifiers`
20150911_130011_emacs exe

**Comment by mbertheau** [comment]
@ralesi: I see it now, thanks :)

I'll try with `color-identifiers` for a while and report back.

**Comment by mbertheau** [comment]
@ralesi Can you add something such that color-identifiers can be toggled on globally for all new buffers?

**Comment by Devagamster** [comment]
```(add-hook 'prog-mode-hook (lambda () (color-identifiers-mode t)))```
:wink:

**Comment by kRITZCREEK** [comment]
Does this only work for elisp or understands color-identifiers a notion of "variables" in other languages too?

**Comment by ralesi** [comment]
@kRItZCREEk, it supports Currently it supports Scala (scala-mode2), JavaScript (js-mode and js2-mode), Ruby, Python, Emacs Lisp, Clojure, C, C++, and Java out of the box.

**Comment by kRITZCREEK** [comment]
@ralesi so does it still provide colored identifiers for other languages? The docs don't say anything about that. And that would be quite the regression from rainbow-identifiers.

**Comment by Devagamster** [comment]
We could have both packages and use color identifiers for the supported languages and then rainbow identifiers for the rest. Best of both worlds

Sent from my Windows Phone
________________________________

**Comment by robbyoconnor** [comment]
<3 <3

**Comment by ralesi** [comment]
We can also modify `color-identifiers:modes-alist` to support  the missing modes.

**Comment by syl20bnr** [comment]
Thank you ! 👍 
I made the choice between color-identifiers and rainbow-identifiers available via the new layer variable `colors-colorize-identifiers` which can take the value `variables` (color-identifiers) `all
Cherry-picked into develop branch, you can safely delete your branch.

**Pull request closed without merging**